Refactor: separate wallet from node

ryanofsky opened the pull request:
This PR is the last in a chain of PRs (#14437, #14711, and #15288) that make the wallet code access node state through an abstract [`Chain`](https://github.com/ryanofsky/bitcoin/blob/pr/wipc-sep/src/interfaces/chain.h) class in [`src/interfaces/`](https://github.com/ryanofsky/bitcoin/tree/pr/wipc-sep/src/interfaces) instead of using global variables like `cs_main`, `chainActive`, and `g_connman`. After this PR, wallet code no longer accesses global variables declared outside the wallet directory, and no longer calls functions accessing those globals (as verified by the `hide-globals` script in #10244).

This PR and the previous PRs have been refactoring changes that do not affect behavior. Previous PRs have consisted of lots of mechanical changes like:

```diff
-    wtx.nTimeReceived = GetAdjustedTime();
+    wtx.nTimeReceived = m_chain->getAdjustedTime();
```

This PR is smaller, but less mechanical. It replaces last few bits of wallet code that access node state directly (through `CValidationInterface`, `CRPCTable`, and `CCoinsViewMemPool` interfaces) with code that uses the `Chain` interface.

These changes allow followup PR #10102 (multiprocess gui & wallet PR) to work without any significant updates to wallet code. Additionally they:

* Provide a single place to describe the interface between wallet and node code.
* Can make better wallet testing possible, because the `Chain` object consists of virtual methods that can be overloaded for mocking. (This could be used to test edge cases in the rescan code, for example).

ryanofsky committed:
Remove use CValidationInterface in wallet code

This commit does not change behavior.

ryanofsky committed:
Remove use of CRPCTable::appendCommand in wallet code

This commit does not change behavior.

corebob commented:
Somewhat related (I think), having the fee rate defined as a class causes some really annoying workarounds, in particular the UniValueType wrapper in rpc/server.h

It seems to me that declaring fee rate as a type that is compatible with UniValue, like CAmount, could avoid some unnecessary cruft.

Anyone have any thoughts on that?

jonasschnelli commented:
Concept ACK

ryanofsky commented:
@jnewbery, the change to stop deduping link arguments is in Makefile.am [here](https://github.com/bitcoin/bitcoin/pull/10973/commits/7bfb409ad9e27df07108c17d63c3efc5f4bfb6a6)

JeremyRubin commented on src/init.cpp:
```diff
@@ -887,7 +888,7 @@ bool AppInitBasicSetup()
 }
 
-bool AppInitParameterInteraction()
+bool AppInitParameterInteraction(ipc::Chain& ipc_chain, ipc::Chain::Clients& ipc_clients)
```
Can we have something that wraps both ipc_chain and ipc_clients for passing in? Is there a reason to have them separate? I don't want to add another useless class, it just seems they are always passed around together & are created together.

JeremyRubin commented on src/ipc/interfaces.h:
```diff
@@ -14,6 +14,26 @@ class Chain
+    //! the LockedState interface and instead call higher-level Chain methods
+    //! that return more information so the chain doesn't need to stay locked
+    //! between calls.
+    class LockedState
```
nit: semantics of name LockedState is maybe a bit clearer as ChainStateLock?

JeremyRubin reviewed (commented):
I did a superficial review of the code & the approach seems reasonable to me!

utACKs from me for the commits checked below...

- [X]    40afc26350 Add skeleton chain and client classes
- [X]    5269b56ec4 Pass chain and client variables where needed
- [X]    7bfb409ad9 Remove uses of wallet functions in init.cpp
- [X]    e26e00bd3f Remove uses of cs_main in wallet code
- [X]    0a3464bce7 Pass chain locked variables where needed
- [X]    c1194635c6 Remove uses of chainActive and mapBlockIndex in wallet code
- [X]    2dd492b580 Remove uses of CheckFinalTx in wallet code
- [X]   ebb23cbe04 Remove use of IsWitnessEnabled in wallet code
- [X]    dd180da24b Remove use of AcceptToMemoryPool in wallet code
- [X]    83be2e6552 Remove uses of GetVirtualTransactionSize in wallet code
- [X]    8931629480 Remove use of IsRBFOptIn in wallet code
- [ ]    684e807240 Remove use of GetCountWithDescendants in wallet code
- [ ]    f3f36e8b5f Remove use of g_connman / PushInventory in wallet code
- [ ]   c9d049c7e4 Remove use of TransactionWithinChainLimit in wallet code
- [ ]    04f6a9ecc1 Remove use of CalculateMemPoolAncestors in wallet code
- [ ]    c510dde004 Remove uses of fee globals in wallet code
- [ ]    9849c4d441 Remove uses of fPruneMode in wallet code
- [ ]    ab488c8229 Remove uses of g_connman in wallet code
- [ ]    090411ad7c Remove uses of GetAdjustedTime in wallet code
- [ ]    64ad91dbf5 Remove uses of InitMessage/Warning/Error in wallet code
- [ ]    3ab3335d38 Remove use CValidationInterface in wallet code
- [ ]    8756810d58 Remove use of CRPCTable::appendCommand in wallet code
- [ ]    35bfb7f189 Remove use of generateBlocks in wallet code
- [ ]    8fb011e3fb Remove uses of ParseConfirmTarget in wallet code

MarcoFalke commented on src/interfaces/chain.cpp:
```diff
@@ -0,0 +1,438 @@
+#include <algorithm>
+#include <future>
+#include <map>
+#include <memory>
```
```
Include(s) from src/interfaces/chain.h duplicated in src/interfaces/chain.cpp:
#include <memory>
Include(s) from src/interfaces/wallet.h duplicated in src/interfaces/wallet.cpp:
#include <memory>
#include <string>
#include <utility>
#include <vector>
Include(s) from src/wallet/init.h duplicated in src/wallet/init.cpp:
#include <walletinitinterface.h>
^---- failure generated from contrib/devtools/lint-includes.sh

ryanofsky replied:
> failure generated from contrib/devtools/lint-includes.sh

I'll try to submit a pr to simplify the developer guideline and this lint check. This check is incompatible with the [iwyu tool](https://github.com/include-what-you-use/include-what-you-use/blob/master/docs/WhyIWYU.md), and I think "include what you use" is a better and simpler policy than "include what you use most of the time but not if the same include line appears in a different file"

MarcoFalke replied:
@ryanofsky Agree with you here. I merged that pull request because it had some acks and no nack, afaics. I am happy to ACK a pull that adjusts the developer guidelines and removes the lint check.

ryanofsky commented:
@skeees, this PR might be relevant to the `GetDepthInMainChain` discussion from https://github.com/bitcoin/bitcoin/pull/12801#issuecomment-379930111. This PR replaces direct accesses to `chainActive` and `cs_main`  global variables in the wallet with calls to a `Chain::Lock` interface:

https://github.com/bitcoin/bitcoin/blob/6efd1524caf008641c4ffc15e8a2b2c2586c6d0f/src/interfaces/chain.h#L42-L111

If you look at [wallet.cpp](https://github.com/ryanofsky/bitcoin/blob/pr/wipc-sep/src/wallet/wallet.cpp) and search for `locked_chain`,  you can see all the things the wallet is currently doing while assuming the chain is locked.

Ideally we would like to eliminate these `locked_chain` objects, or make them very rarely used. I think the single change that would eliminate most uses of `locked_chain` would be to add an `m_last_block_processed_height` member to CWallet and an `m_block_height` member to `CMerkleTx`. These variables could get updated in `BlockConnected`, so methods like `GetDepthInMainChain`, `IsInMainChain`, and `GetBlocksToMaturity` could be implemented without external locking.

This is a change which I think could be implemented right now in a new PR that doesn't depend on anything. And there are probably other similar changes which could eliminate more locking and which your #12801 notifier might come in handy for.

jamesob commented:
Needs rebase.

ryanofsky commented:
> Needs rebase.

Rebased ce6af5dd9e70db2dfcd5d5f1ae10494b5129ff88 -> af8f8087699c30ec83995c54112955253bcaba84 ([pr/wipc-sep.50](https://github.com/ryanofsky/bitcoin/commits/pr/wipc-sep.50) -> [pr/wipc-sep.51](https://github.com/ryanofsky/bitcoin/commits/pr/wipc-sep.51)) due to conflicts with #12639 and #12507.

(Reminders to rebase this are welcome, but not needed. Since this conflicts with most wallet prs, I tend to check this frequently and rebase it a few times per week).

jamesob reviewed (commented):
utACK https://github.com/bitcoin/bitcoin/pull/10973/commits/af8f8087699c30ec83995c54112955253bcaba84

I completed a commit-by-commit readthrough of the code and don't have any substantial feedback aside from a few typos.

Reviewers will notice that dealing in the more primitive datatypes returned by `chain` methods (block heights and hashes as opposed to, say, full `CBlockIndex` values) results in more fiddling at callsites (null checks, arithmetic, subsequent lookups, etc.). This is an obvious drawback, but if we imagine that these calls are happening over a process boundary - as, god willing, they someday will - then the simpler the datatype the better. The stripped-down return types also more tightly accommodate the data requirements of the callers. For these two reasons, this changeset is a big step towards system decomposition IMO.

Small additional notes:
- there are some lingering references to `chainActive` in `wallet/test/wallet_tests.cpp`; not sure if we want to remove those or not.
- there's a forward declaration of `CBlockIndex` in `wallet/wallet.h` that doesn't look necessary anymore.

I'll be doing a manual test of this branch in the next few days.

jnewbery commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2214,39 +2227,41 @@ UniValue listsinceblock(const JSONRPCRequest& request)
 
-    int depth = pindex ? (1 + chainActive.Height() - pindex->nHeight) : -1;
+    const Optional<int> tip_height = locked_chain->getHeight();
+    int depth = tip_height && height ? (1 + *tip_height - *height) : -1;
```
Can you use `locked_chain->getDepth(blockId)` here?

jamesob replied:
We need `tip_height` for later and so using `getDepth()` would do an extra `getHeight()` call. But agree that it's a shame to see the depth arithmetic repeated...

jnewbery reviewed (commented):
I've reviewed the first 5 commits, and everything in *Remove uses of chainActive and mapBlockIndex in wallet code* except for wallet.cpp. Will continue on Monday.

A few questions/comments inline.

jnewbery commented on src/interfaces/chain.h:
```diff
@@ -0,0 +1,251 @@
+        virtual Optional<int> findEarliestAtLeast(int64_t time) = 0;
+
+        //! Return height of last block in chain with timestamp less than the
+        //! given, and height less than or equal to the given, or nothing if
```
Shouldn't this be "height greater than or equal to the given"? The implementation starts at `start_height` and moves forwards along the chain.

jnewbery reviewed (commented):
Finished reviewing _Remove uses of chainActive and mapBlockIndex in wallet code_. A few more comments inline.

jnewbery commented on src/interfaces/chain.cpp:
```diff
@@ -0,0 +1,419 @@
+
+    UniValue forwardRequest(const JSONRPCRequest& request) const
+    {
+        // Simple forwarding of RPC requests. This just sends the request to the
```
Supporting multiple clients makes this more complex than it needs to be.

This PR would be easier to review if it initially supported just a single client (as we do today). A future PR could extend the RPC forwarding to support multiple clients.

jnewbery reviewed (commented):
Finished initial review of all commits. One more comment inline that I think the RPC forwarding code would be a lot easier to review if it was limited to a single client. Future PRs could add support for multiple clients.

jnewbery commented:
Overall, I think this is a huge improvement. It gives us a well-defined interface between the wallet and node and is a huge step towards separating out the different subsystems. Whether or not we decide to go ahead with process separation, this is very useful work.

The wallet<->node interface could certainly be tidied up in future PRs. Since this is an internal interface, that shouldn't stop this PR from being merged as the first step.

ryanofsky reviewed (commented):
Thanks for the reviews!

Rebased af8f8087699c30ec83995c54112955253bcaba84 -> 1c597aab8749c95f6ceef24d6795781c913f74a3 ([pr/wipc-sep.51](https://github.com/ryanofsky/bitcoin/commits/pr/wipc-sep.51) -> [pr/wipc-sep.52](https://github.com/ryanofsky/bitcoin/commits/pr/wipc-sep.52)) due to conflicts with #13163 and #13079
Added 15 commits 1c597aab8749c95f6ceef24d6795781c913f74a3 -> 46593b55552f40dfbb72868c822aafd9834e45ee ([pr/wipc-sep.52](https://github.com/ryanofsky/bitcoin/commits/pr/wipc-sep.52) -> [pr/wipc-sep.53](https://github.com/ryanofsky/bitcoin/commits/pr/wipc-sep.53), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/wipc-sep.52...pr/wipc-sep.53))
Squashed 46593b55552f40dfbb72868c822aafd9834e45ee -> 1bf534a9f53e5ba6f12f05f5e18739715c821576 ([pr/wipc-sep.53](https://github.com/ryanofsky/bitcoin/commits/pr/wipc-sep.53) -> [pr/wipc-sep.54](https://github.com/ryanofsky/bitcoin/commits/pr/wipc-sep.54))
Rebased 1bf534a9f53e5ba6f12f05f5e18739715c821576 -> b13720a225f50f9fa96b7e10fda181b1d2770cb8 ([pr/wipc-sep.54](https://github.com/ryanofsky/bitcoin/commits/pr/wipc-sep.54) -> [pr/wipc-sep.55](https://github.com/ryanofsky/bitcoin/commits/pr/wipc-sep.55)) due to conflicts with #13081 and #12560

sipa commented:
Concept ACK on a better defined interface between wallet and node.

promag commented:
Concept ACK, makes sense!

jonasschnelli commented:
Concept ACK

jnewbery commented:
Great. 3 concept ACKs!

Splitting first 6 commits into their own PR could make this easier to review: https://botbot.me/freenode/bitcoin-core-dev/2018-05-17/?msg=100176268&page=4

@ryanofsky - let me know if I can help with that.

ryanofsky commented:
Rebased b13720a225f50f9fa96b7e10fda181b1d2770cb8 -> a71c5b8e73d991f28945280c8812fa0c2898a710 ([pr/wipc-sep.55](https://github.com/ryanofsky/bitcoin/commits/pr/wipc-sep.55) -> [pr/wipc-sep.56](https://github.com/ryanofsky/bitcoin/commits/pr/wipc-sep.56)) due to conflict with #10740

jimpo commented on src/interfaces/chain.h:
```diff
@@ -182,6 +182,9 @@ class Chain
         bool keep_script) = 0;
 
+    //! Parse confirm target.
+    virtual unsigned int parseConfirmTarget(const UniValue& value) = 0;
```
This feels like a weird method to put in the chain interface because it's just an RPC helper function. I'd actually rather it be duplicated (though that might still require a `maxConfirmTarget` method on the interface).

MarcoFalke commented:
utACK f06a3cc9190bcac4f6109bf9a9f03f92edf1d1fd (not the head commit)

Will leave two questions for clarification

MarcoFalke commented on src/wallet/rpcwallet.cpp:
```diff
@@ -66,11 +66,6 @@ bool EnsureWalletIsAvailable(CWallet * const pwallet, bool avoidException)
-        // wallet RPC methods aren't registered unless a wallet is loaded. But
-        // this error is being kept as a precaution, because it's possible in
-        // the future that wallet RPC methods might get or remain registered
-        // when no wallets are loaded.
```
Mind to explain why this is removed? I believe `Construct` will still return without creating a chain client and thus `RegisterWalletRPCCommands` is never called.

ken2812221 commented on src/optional.h:
```diff
@@ -0,0 +1,18 @@
+#define BITCOIN_OPTIONAL_H
+
+#include <boost/none.hpp>
+#include <boost/optional/optional.hpp>
```
Maybe replace these two lines as `#include <boost/optional.hpp>` to pass travis tests.

In commit 'Remove uses of chainActive and mapBlockIndex in wallet code '

jnewbery commented:
@ryanofsky - are you still maintaining this? What are your thoughts about splitting the first 6 commits into a separate PR (https://github.com/bitcoin/bitcoin/pull/10973#issuecomment-389991584)?

ryanofsky committed:
Remove use of CCoinsViewMemPool::GetCoin in wallet code

This commit does not change behavior.

ryanofsky commented:
> Needs rebase

Rebased 7e59fa372968009d8b4df29ce9893b4038f9c2cd -> 7f9fabce8fa33aae220f9cdeb40c540ef0ef849d ([pr/wipc-sep.62](https://github.com/ryanofsky/bitcoin/commits/pr/wipc-sep.62) -> [pr/wipc-sep.63](https://github.com/ryanofsky/bitcoin/commits/pr/wipc-sep.63)) due to conflict with #13235.
Rebased 7f9fabce8fa33aae220f9cdeb40c540ef0ef849d -> 56e391398edc617042d910a2433ccd08c6a05d44 ([pr/wipc-sep.63](https://github.com/ryanofsky/bitcoin/commits/pr/wipc-sep.63) -> [pr/wipc-sep.64](https://github.com/ryanofsky/bitcoin/commits/pr/wipc-sep.64)) due to conflict with #13570.
Rebased 56e391398edc617042d910a2433ccd08c6a05d44 -> d8abd3c2a9378968d258b8beb33b6516eb2257fa ([pr/wipc-sep.64](https://github.com/ryanofsky/bitcoin/commits/pr/wipc-sep.64) -> [pr/wipc-sep.65](https://github.com/ryanofsky/bitcoin/commits/pr/wipc-sep.65)) due to conflict with #13622.
Rebased d8abd3c2a9378968d258b8beb33b6516eb2257fa -> 635bc2d32ea2a5e67bef8669cc4b4466489d72d1 ([pr/wipc-sep.65](https://github.com/ryanofsky/bitcoin/commits/pr/wipc-sep.65) -> [pr/wipc-sep.66](https://github.com/ryanofsky/bitcoin/commits/pr/wipc-sep.66)) due to conflicts with #13630, #13566, #13651, #13072, and #12944. Also added `g_interfaces` global variable to be able to access chain and chain interface pointers from RPC handlers like `setmocktime` and `loadwallet` that need access to the interfaces but don't have any other way of getting access to bitcoin state.
Rebased 635bc2d32ea2a5e67bef8669cc4b4466489d72d1 -> 3344aa65fc6179dabc6fa736fe6c81043466f221 ([pr/wipc-sep.66](https://github.com/ryanofsky/bitcoin/commits/pr/wipc-sep.66) -> [pr/wipc-sep.67](https://github.com/ryanofsky/bitcoin/commits/pr/wipc-sep.67)) due to conflicts with #13822, #9662, #13786. Also replaced ` g_interfaces` with `g_rpc_interfaces`.
Updated 3344aa65fc6179dabc6fa736fe6c81043466f221 -> fb60d4fbbe5f5e588618a41b858e376f938a02a5 ([pr/wipc-sep.67](https://github.com/ryanofsky/bitcoin/commits/pr/wipc-sep.67) -> [pr/wipc-sep.68](https://github.com/ryanofsky/bitcoin/commits/pr/wipc-sep.68)) to work around circular dependency lint error.

jamesob commented on src/wallet/wallet.cpp:
```diff
@@ -2726,7 +2741,8 @@ bool CWallet::CreateTransaction(interfaces::Chain::Lock& locked_chain, const std
     // nLockTime that preclude a fix later.
-    txNew.nLockTime = chainActive.Height();
+    const Optional<int> tip_height = locked_chain.getHeight();
+    txNew.nLockTime = tip_height ? *tip_height : std::numeric_limits<uint32_t>::max();
```
Probably out-of-scope nit: I wish the uint32_t max was instead a named constant that had meaning relative to use with nLockTime (e.g. `IGNORE_NLOCKTIME`).

ken2812221 commented:
- [x] 64e226b6fe Add skeleton chain and client classes
- [ ] d9f54ba5fa Pass chain and client variables where needed
- [ ] 20887df606 Remove direct node->wallet calls in init.cpp
- [ ] 1635206668 Remove uses of cs_main in wallet code
- [ ] 283e874fd7 Pass chain locked variables where needed
- [ ] cb0c6b42a0 Remove uses of chainActive and mapBlockIndex in wallet code
- [ ] 926c7d2f40 Remove uses of CheckFinalTx in wallet code
- [ ] 0c3483559d Remove use of AcceptToMemoryPool in wallet code
- [ ] 627e13a450 Remove uses of GetVirtualTransactionSize in wallet code
- [ ] ec46a2ccfb Remove use of IsRBFOptIn in wallet code
- [ ] 151d6cb750 Remove use of GetCountWithDescendants in wallet code
- [ ] 31591bf12c Remove use of g_connman / PushInventory in wallet code
- [ ] 5db710cdb8 Remove use of GetTransactionAncestry in wallet code
- [ ] 1ca7a0447d Remove use of CalculateMemPoolAncestors in wallet code
- [ ] 55a179e263 Remove uses of fee globals in wallet code
- [ ] 5eacb1ea62 Remove uses of fPruneMode in wallet code
- [ ] 0902deb3eb Remove uses of g_connman in wallet code
- [ ] 6b32c54269 Remove uses of GetAdjustedTime in wallet code
- [ ] fa52ad7530 Remove uses of InitMessage/Warning/Error in wallet code
- [ ] 8ac1cb6397 Remove use CValidationInterface in wallet code
- [ ] 84cf3a3ffd Remove use of CRPCTable::appendCommand in wallet code
- [ ] 9c2b47adc4 Remove use of generateBlocks in wallet code
- [ ] e8477b1c5b Remove uses of ParseConfirmTarget in wallet code
- [ ] 5e43ea8f43 Remove use of uiInterface.LoadWallet in wallet code
- [ ] 2019b6a1ac Remove use of IsInitialBlockDownload in wallet code
- [ ] 735b242930 Remove use of StartShutdown in wallet code
- [ ] 249bf50061 Remove use of CCoinsViewMemPool::GetCoin in wallet code

laanwj commented:
Concept ACK

ken2812221 reviewed (approved):
utACK e44e639558b1084f14a97847592616c3df9fff38

practicalswift commented:
This PR does not seem to compile when rebased on `master` :-)

practicalswift commented on src/wallet/test/wallet_test_fixture.cpp:
```diff
@@ -8,16 +8,15 @@
+    m_chain_client->registerRpcs();
 }
 
 WalletTestingSetup::~WalletTestingSetup()
```
```
2018-09-19 14:39:06 clang-tidy(pr=10973): src/wallet/test/wallet_test_fixture.cpp:20:21: warning: use '= default' to define a trivial destructor [hicpp-use-equals-default]
```

practicalswift commented on src/interfaces/chain.cpp:
```diff
@@ -0,0 +1,453 @@
+    bool haveBlockOnDisk(int height) override
+    {
+        CBlockIndex* block = ::chainActive[height];
+        return block && (block->nStatus & BLOCK_HAVE_DATA) && block->nTx > 0;
```
```
2018-09-19 14:39:06 clang-tidy(pr=10973): src/interfaces/chain.cpp:81:25: warning: implicit conversion 'unsigned int' -> bool [readability-implicit-bool-conversion]
```

practicalswift commented on src/interfaces/chain.cpp:
```diff
@@ -0,0 +1,453 @@
+        if (::fPruneMode) {
+            CBlockIndex* block = stop_height ? ::chainActive[*stop_height] : ::chainActive.Tip();
+            while (block && block->nHeight >= start_height) {
+                if (!(block->nStatus & BLOCK_HAVE_DATA)) {
```
```
2018-09-19 14:39:06 clang-tidy(pr=10973): src/interfaces/chain.cpp:105:22: warning: implicit conversion 'unsigned int' -> bool [readability-implicit-bool-conversion]
```

practicalswift commented on src/interfaces/chain.cpp:
```diff
@@ -0,0 +1,453 @@
+            block = it->second;
+            fork = ::chainActive.FindFork(block);
+        }
+        if (height) {
```
```
2018-09-19 14:39:06 clang-tidy(pr=10973): src/interfaces/chain.cpp:121:13: warning: implicit conversion 'Optional<int> *' (aka 'boost::optional<int> *') -> bool [readability-implicit-bool-conversion]
```

flack replied:
@practicalswift You know, the signal/noise ratio of these comments would be higher if you omitted the timestamp (which isn't interesting) and the file path (which is redundant, because that's where the comment is attached). Currently in the web UI, you always have to scroll each comment horizontally to see what it's actually about.

status quo:
<img width="676" alt="bildschirmfoto 2018-09-23 um 18 40 53" src="https://user-images.githubusercontent.com/425166/45930419-5f36e980-bf60-11e8-866f-102591e58d38.png">

proposed:
<img width="664" alt="bildschirmfoto 2018-09-23 um 18 41 21" src="https://user-images.githubusercontent.com/425166/45930424-6958e800-bf60-11e8-841b-1e553ca9d7b6.png">

practicalswift commented on src/interfaces/chain.cpp:
```diff
@@ -0,0 +1,453 @@
+class HandlerImpl : public Handler
+{
+public:
+    HandlerImpl(Chain::Notifications& notifications) : m_forwarder(notifications) {}
```
```
2018-09-19 14:39:06 clang-tidy(pr=10973): src/interfaces/chain.cpp:168:5: warning: single-argument constructors must be marked explicit to avoid unintentional implicit conversions [hicpp-explicit-conversions]
```

practicalswift commented on src/interfaces/chain.cpp:
```diff
@@ -0,0 +1,453 @@
+
+    struct Forwarder : CValidationInterface
+    {
+        Forwarder(Chain::Notifications& notifications) : m_notifications(&notifications)
```
```
2018-09-19 14:39:06 clang-tidy(pr=10973): src/interfaces/chain.cpp:182:9: warning: single-argument constructors must be marked explicit to avoid unintentional implicit conversions [google-explicit-constructor]
```

practicalswift replied:
```
2018-09-20 04:12:32 cppcheck(pr=10973): [src/interfaces/chain.cpp:182]: (style) Struct 'Forwarder' has a constructor with 1 argument that is not explicit.
```

practicalswift commented on src/interfaces/chain.cpp:
```diff
@@ -0,0 +1,453 @@
+            m_notifications->BlockDisconnected(*block);
+        }
+        void ChainStateFlushed(const CBlockLocator& locator) override { m_notifications->ChainStateFlushed(locator); }
+        void ResendWalletTransactions(int64_t best_block_time, CConnman* connman) override
```
```
2018-09-19 14:39:06 clang-tidy(pr=10973): src/interfaces/chain.cpp:212:74: warning: parameter 'connman' is unused [misc-unused-parameters]
```

practicalswift commented on src/interfaces/chain.cpp:
```diff
@@ -0,0 +1,453 @@
+        LOCK(::mempool.cs);
+        if (!::mempool.CalculateMemPoolAncestors(entry, setAncestors, nLimitAncestors, nLimitAncestorSize,
+                nLimitDescendants, nLimitDescendantSize, errString)) {
+            return false;
```
```
2018-09-19 14:39:06 clang-tidy(pr=10973): src/interfaces/chain.cpp:315:20: warning: redundant boolean literal in conditional return statement [readability-simplify-boolean-expr]
```

practicalswift commented on src/interfaces/chain.cpp:
```diff
@@ -0,0 +1,453 @@
+class ChainImpl::RpcForwarder
+{
+public:
+    RpcForwarder(const CRPCCommand& command) : m_command(command)
```
```
2018-09-19 14:39:06 clang-tidy(pr=10973): src/interfaces/chain.cpp:362:5: warning: single-argument constructors must be marked explicit to avoid unintentional implicit conversions [google-explicit-constructor]
```

practicalswift replied:
```
2018-09-20 04:12:32 cppcheck(pr=10973): [src/interfaces/chain.cpp:362]: (style) Class 'RpcForwarder' has a constructor with 1 argument that is not explicit.
```

practicalswift commented on src/interfaces/chain.cpp:
```diff
@@ -0,0 +1,453 @@
+
+    void removeCommand(const CRPCCommand& command)
+    {
+        m_commands.erase(std::remove(m_commands.begin(), m_commands.end(), &command));
```
```
2018-09-19 14:39:06 clang-tidy(pr=10973): src/interfaces/chain.cpp:373:9: warning: this call will remove at most one item even when multiple items should be removed [bugprone-inaccurate-erase]
```

practicalswift commented on src/interfaces/chain.cpp:
```diff
@@ -0,0 +1,453 @@
+        }
+    }
+
+    ~RpcHandler() { disconnect(); }
```
```
2018-09-19 14:39:06 clang-tidy(pr=10973): src/interfaces/chain.cpp:432:5: warning: annotate this function with 'override' or (rarely) 'final' [hicpp-use-override]
```

practicalswift replied:
```
2018-09-19 14:39:06 clang-tidy(pr=10973): src/interfaces/chain.cpp:432:21: warning: Call to virtual function during destruction [clang-analyzer-optin.cplusplus.VirtualCall]
```

practicalswift commented on src/wallet/wallet.cpp:
```diff
@@ -945,19 +947,19 @@ void CWallet::LoadToWallet(const CWalletTx& wtxIn)
 }
 
-bool CWallet::AddToWalletIfInvolvingMe(const CTransactionRef& ptx, const CBlockIndex* pIndex, int posInBlock, bool fUpdate)
+bool CWallet::AddToWalletIfInvolvingMe(interfaces::Chain::Lock& locked_chain, const CTransactionRef& ptx, const uint256& block_hash, int posInBlock, bool fUpdate)
```
```
2018-09-19 14:39:06 clang-tidy(pr=10973): src/wallet/wallet.cpp:950:65: warning: parameter 'locked_chain' is unused [misc-unused-parameters]
```

practicalswift commented on src/threadsafety.h:
```diff
@@ -54,4 +54,15 @@
+struct SCOPED_LOCKABLE LockAnnotation
+{
+    template <typename Mutex>
+    LockAnnotation(Mutex& mutex) EXCLUSIVE_LOCK_FUNCTION(mutex)
```
```
2018-09-20 04:12:32 cppcheck(pr=10973): [src/threadsafety.h:62]: (style) Struct 'LockAnnotation' has a constructor with 1 argument that is not explicit.
```

practicalswift commented on src/interfaces/chain.h:
```diff
@@ -0,0 +1,265 @@
+        }
+        virtual void BlockDisconnected(const CBlock& block) {}
+        virtual void ChainStateFlushed(const CBlockLocator& locator) {}
+        virtual void Inventory(const uint256& hash) {}
```
```
2018-09-20 04:12:32 cppcheck(pr=10973): [src/interfaces/chain.h:209]: (style) The function 'Inventory' is never used.
```

jamesob commented:
Should this be closed in lieu of https://github.com/bitcoin/bitcoin/pull/14437?

ryanofsky commented:
> Should this be closed now that #14437 has been opened?

I updated the PR description above to make it obvious that review comments be added in #14437. I'd still like to keep updating this PR though. I think for example being able to see the complete [`Chain`](https://github.com/ryanofsky/bitcoin/blob/pr/wipc-sep/src/interfaces/chain.h) interface added here is useful for understanding the direction #14437 is starting to go in.

practicalswift commented on src/interfaces/chain.cpp:
```diff
@@ -0,0 +1,454 @@
+        // will keep sending requests here even if there are no clients left to
+        // forward to.)
+        throw JSONRPCError(RPC_METHOD_NOT_FOUND, "Method not found");
+    };
```
Nit: Remove `;` :-)

practicalswift commented on src/rpc/rawtransaction.cpp:
```diff
@@ -754,22 +757,44 @@ static UniValue combinerawtransaction(const JSONRPCRequest& request)
+        m_cache.SetBackend(*this);
+        m_cache.AccessCoin(output);
+    }
+    ~CoinFill() {
```
Should be marked `override`?

jnewbery commented:
@practicalswift - see note in PR description:

> This PR is based on #14437, so review comments for initial commits should be posted there.

I recommend you don't spend time reviewing this until #14437 is merged.

practicalswift commented on src/interfaces/chain.cpp:
```diff
@@ -0,0 +1,454 @@
+        LockPoints lp;
+        CTxMemPoolEntry entry(tx, 0, 0, 0, false, 0, lp);
+        CTxMemPool::setEntries setAncestors;
+        size_t nLimitAncestors = gArgs.GetArg("-limitancestorcount", DEFAULT_ANCESTOR_LIMIT);
```
Make this and the conversions on the lines below explicit?

practicalswift commented on src/wallet/fees.cpp:
```diff
@@ -90,10 +89,10 @@ CFeeRate GetMinimumFeeRate(const CWallet& wallet, const CCoinControl& coin_contr
 {
-    unsigned int highest_target = estimator.HighestTargetTracked(FeeEstimateHorizon::LONG_HALFLIFE);
-    CFeeRate discard_rate = estimator.estimateSmartFee(highest_target, nullptr /* FeeCalculation */, false /* conservative */);
+    unsigned int highest_target = wallet.chain().estimateMaxBlocks();
```
Implicit sign conversion here: see comment on `estimateMaxBlocks` above.

ryanofsky reviewed (commented):
re: https://github.com/bitcoin/bitcoin/pull/10973#pullrequestreview-116397255
> there are some lingering references to chainActive in wallet/test/wallet_tests.cpp

Maybe some of these could be changed, but I think it's fine if test code manipulates chain state directly. Alternatives would be extending the `Chain` interface or writing a new interface to manipulate chain state, which would add complexity and overhead, or having tests create mock chain objects, which might make sense but could result in more test code and less realistic tests.

re: https://github.com/bitcoin/bitcoin/pull/10973#pullrequestreview-116397255
> there's a forward declaration of CBlockIndex in wallet/wallet.h that doesn't look necessary anymore.

Thanks, removed this.

re: https://github.com/bitcoin/bitcoin/pull/10973#issuecomment-390335988
> Also, I'll note that the lowercasing of method names on the interfaces conflicts with the Coding Style.

Thanks, sipa also commented on this in https://github.com/bitcoin/bitcoin/pull/14437#issuecomment-429439919. Will try a PR to update the style.

practicalswift commented on src/init.cpp:
```diff
@@ -1243,7 +1258,11 @@ bool AppInitMain()
 
     // ********************************************************* Step 5: verify wallet database integrity
-    if (!g_wallet_init_interface.Verify()) return false;
+    for (const auto& client : interfaces.chain_clients) {
```
Could use `std::any_of`? :-)

practicalswift commented on src/init.cpp:
```diff
@@ -1562,7 +1581,11 @@ bool AppInitMain()
 
     // ********************************************************* Step 9: load wallet
-    if (!g_wallet_init_interface.Open()) return false;
+    for (const auto& client : interfaces.chain_clients) {
```
Could use `std::any_of`? :-)

practicalswift commented on src/init.cpp:
```diff
@@ -1243,7 +1258,11 @@ bool AppInitMain()
-    if (!g_wallet_init_interface.Verify()) return false;
+    for (const auto& client : interfaces.chain_clients) {
+        if (!client->verify()) {
+            return false;
```
Nit: Looks like a potential opportunity for `std::any_of`? :-)

practicalswift commented on src/init.cpp:
```diff
@@ -1562,7 +1581,11 @@ bool AppInitMain()
-    if (!g_wallet_init_interface.Open()) return false;
+    for (const auto& client : interfaces.chain_clients) {
+        if (!client->load()) {
+            return false;
```
Nit: Looks like a potential opportunity for `std::any_of`? :-)

MeshCollider commented:
Concept ACK, will this PR be rebased and ready for review now #14437 has been merged, or will you make another smaller PR for the next steps and leave this just to track overall?

ryanofsky reviewed (commented):
> will this PR be rebased and ready for review now #14437 has been merged, or will you make another smaller PR for the next steps and leave this just to track overall?

I made a new PR #14711 containing the first commit from this PR, but I'm also maintaining this.

MarcoFalke commented:
utACK 6898f53760

ryanofsky committed:
Remove remaining wallet accesses to node globals

promag commented on src/interfaces/chain.h:
```diff
@@ -160,6 +163,9 @@ class Chain
         int64_t* max_time = nullptr) = 0;
 
+    //! Get unspent outputs associated with a transaction.
+    virtual std::vector<Coin> findCoins(const std::vector<COutPoint>& outputs) = 0;
```
I wonder if this should return an `std::vector<std::pair<COutPoint, Coin>>`? Otherwise could note in the comment that result follows `outputs` order.

promag commented on src/rpc/rawtransaction.cpp:
```diff
@@ -790,22 +791,44 @@ static UniValue combinerawtransaction(const JSONRPCRequest& request)
+    }
+    CCoinsViewCache& m_cache;
+    const COutPoint& m_output;
+    Coin&& m_coin;
```
Why?

promag commented on src/rpc/server.cpp:
```diff
@@ -193,9 +193,8 @@ std::string CRPCTable::help(const std::string& strCommand, const JSONRPCRequest&
-            rpcfn_type pfn = pcmd->actor;
-            if (setDone.insert(pfn).second)
-                (*pfn)(jreq);
+            if (setDone.insert(pcmd->unique_id).second)
```
nit, `{`

promag commented on src/rpc/rawtransaction.cpp:
```diff
@@ -790,22 +791,44 @@ static UniValue combinerawtransaction(const JSONRPCRequest& request)
+        }
+        std::vector<Coin> coins = chain.findCoins(outputs);
+        for (size_t i = 0; i < coins.size(); ++i) {
+            CoinFill(view, outputs[i], std::move(coins[i]), viewDummy);
```
Can you explain why not do the following:
```cpp
CoinFill view(...);
for (...) {
   view.AccessCoin(...)
}
```

promag commented on src/interfaces/chain.cpp:
```diff
@@ -254,8 +343,111 @@ class ChainImpl : public Chain
+{
+    auto inserted = m_rpc_forwarders.emplace(
+        std::piecewise_construct, std::forward_as_tuple(command.name), std::forward_as_tuple(command));
+    if (inserted.second && !inserted.first->second.registerForwarder()) {
```
When can `inserted.second` fail?

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -4156,8 +4156,8 @@ static const CRPCCommand commands[] =
 {
     for (unsigned int vcidx = 0; vcidx < ARRAYLEN(commands); vcidx++)
-        t.appendCommand(commands[vcidx].name, &commands[vcidx]);
+        handlers.emplace_back(chain.handleRpc(commands[vcidx]));
```
`handleRpc` can return an "empty" handler.

promag replied:
But why let `handlers` have a null entry?

ryanofsky replied:
> But why let handlers have a null entry?

This won't happen and it wouldn't be a problem if it did happen. I'm trying not to change behavior and this is a straight translation of the previous code. Neither the old code nor the new code cares about this condition. If there are concerns about rpc registration they would probably be better to address in another PR.

promag commented on src/interfaces/chain.cpp:
```diff
@@ -254,8 +343,111 @@ class ChainImpl : public Chain
+        return MakeUnique<HandlerImpl>(notifications);
+    }
+    void waitForNotifications() override { SyncWithValidationInterfaceQueue(); }
+    std::unique_ptr<Handler> handleRpc(const CRPCCommand& command) override;
```
nit, maybe a bit late, but should be uppercase RPC?

promag reviewed (commented):
Looks good, as @jnewbery points out, a bit hard to review, but I think it's worth the effort. Sorry for these nits, feel free to ignore them.

> After this PR, wallet code no longer accesses global variables declared outside the wallet directory, and no longer calls functions accessing those globals (as verified by the hide-globals script in #10244).

Do you think this should be committed and verified in CI?

ryanofsky reviewed (commented):
Thanks for reviews!

Updated 6898f5376071dcd84ce4aea980e170b4953c18bb -> 244edfe470d8b6a4c449649952bb2130e0d10bd5 ([pr/wipc-sep.97](https://github.com/ryanofsky/bitcoin/commits/pr/wipc-sep.97) -> [pr/wipc-sep.98](https://github.com/ryanofsky/bitcoin/commits/pr/wipc-sep.98), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/wipc-sep.97..pr/wipc-sep.98)).

* Main change is adding a new commit "Remove remaining wallet accesses to node globals" (244edfe470d8b6a4c449649952bb2130e0d10bd5).
* Commit "Remove use CValidationInterface in wallet code" (5c7f4118ef82248f50ff112828638e0a99247cd9, formerly fce0e3cfd1a8616e1c2c05052af277ae8cd7a574) is simplified to remove workaround for UnregisterValidationInterface crash fixed by 2196c51821e340c9a9d2c76c30f9402370f84994 in #13743.
* Commit "Remove use of CRPCTable::appendCommand in wallet code" (a15a7bf8093c9841aed2e51a6655fc5be5b7d783, formerly 53a730ab0c37354db1333f436761fb81f94e3657) is unchanged except for a code comment.
* Commit "Remove use of CCoinsViewMemPool::GetCoin in wallet code" (9863ddc7dbec47d57967544979f694e740d83916 formerly 6898f5376071dcd84ce4aea980e170b4953c18bb) has been simplified by replacing a custom `CCoinsView` subclass with a plain `std::map`.

re: https://github.com/bitcoin/bitcoin/pull/10973#pullrequestreview-212761964

> > After this PR, wallet code no longer accesses global variables declared outside the wallet directory, and no longer calls functions accessing those globals (as verified by the hide-globals script in #10244).

> Do you think this should be committed and verified in CI?

I'd like to verify this, but not with a script, since it would be awkward to set up and maintain. Instead, I'd like to just add a new `libbitcoin_node.a` library, analogous to the existing `libbitcoin_wallet.a` library, and move symbol definitions like `cs_main` `pcoinsTip` `chainActive` there that wallet and gui code shouldn't access. That way any bad references will just result in link errors while building `bitcoin-wallet` and `bitcoin-gui` executables.

Since we have src/node/ src/wallet/ and src/qt/ folders we could also set up rules preventing source files in one of these folders including headers in the other folders. This would provide the clearest separation, but would also require moving a bunch of code around, so would not be a short-term goal.

ryanofsky commented:
Updated 244edfe470d8b6a4c449649952bb2130e0d10bd5 -> 591434b72c3b0ec218fbe30969ae5a819ca224c4 ([pr/wipc-sep.98](https://github.com/ryanofsky/bitcoin/commits/pr/wipc-sep.98) -> [pr/wipc-sep.99](https://github.com/ryanofsky/bitcoin/commits/pr/wipc-sep.99), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/wipc-sep.98..pr/wipc-sep.99)) to fix travis error with old boost.

MarcoFalke commented:
nit: In commit  591434b

Could also update the TODO comment for `GetAvailableCredit` and remove the `cs_main` from it?

ryanofsky commented:
Updated 591434b72c3b0ec218fbe30969ae5a819ca224c4 -> b7c78217ca7aa861b58fddcfdb82a09ca56f8023 ([pr/wipc-sep.99](https://github.com/ryanofsky/bitcoin/commits/pr/wipc-sep.99) -> [pr/wipc-sep.100](https://github.com/ryanofsky/bitcoin/commits/pr/wipc-sep.100), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/wipc-sep.99..pr/wipc-sep.100)) with Marco's GetAvailableCredit suggestion from https://github.com/bitcoin/bitcoin/pull/10973#issuecomment-472069472

promag commented:
> Instead, I'd like to just add a new `libbitcoin_node.a` library, analogous to the existing `libbitcoin_wallet.a` library, and move symbol definitions like `cs_main` `pcoinsTip` `chainActive` there that wallet and gui code shouldn't access.

Nice!

jnewbery commented on src/interfaces/chain.h:
```diff
@@ -217,6 +236,36 @@ class Chain
+        virtual ~Notifications() {}
+        virtual void TransactionAddedToMempool(const CTransactionRef& tx) {}
+        virtual void TransactionRemovedFromMempool(const CTransactionRef& ptx) {}
+        virtual void BlockConnected(const CBlock& block,
```
Can you simplify this interface to `(cont CBlock& block, const std::vector<CTransactionRef>& tx_conflicted)`? The client can get the block hash using `CBlock->GetHash()`.

jnewbery commented on src/wallet/wallet.cpp:
```diff
@@ -4386,7 +4384,7 @@ std::shared_ptr<CWallet> CWallet::CreateWalletFromFile(interfaces::Chain& chain,
     chain.loadWallet(interfaces::MakeWallet(walletInstance));
 
     // Register with the validation interface. It's ok to do this after rescan since we're still holding cs_main.
```
I think this comment could be changed slightly now, since 'we' (the wallet) don't ever hold `cs_main` now.

`since we're still holding cs_main` -> `since we're still holding the Chain::Lock interface`.

jnewbery commented on src/wallet/wallet.h:
```diff
@@ -808,6 +809,9 @@ class CWallet final : public CCryptoKeyStore, public CValidationInterface
     std::set<COutPoint> setLockedCoins GUARDED_BY(cs_wallet);
 
+    /** Validation event callback handler. */
+    std::unique_ptr<interfaces::Handler> m_handler;
```
I don't really like this name `m_handler`, (or the class name `HandlerImpl`), since the name handler is generic for any interface handler. It's more verbose, but I think `m_validation_interface_handler` and `ValidationInterfaceHandlerImpl` are more explicit.

jnewbery reviewed (commented):
I've reviewed 5c7f4118ef82248f50ff112828638e0a99247cd9 (_Remove use CValidationInterface in wallet code_) and it looks good so far! A few comments inline.

Incidentally, this github PR is pretty unwieldly - the notes are very long, and without reading through them all it's difficult to tell which comments are for the overall approach and which are for the commits under review now. I don't know if's worth it at this point, but it might be clearer to open one final PR for the last commits, and close this or leave it as the overarching PR.

ryanofsky reviewed (commented):
Updated b7c78217ca7aa861b58fddcfdb82a09ca56f8023 -> 1a1036a5b3bc48715a5f3955e87203287c37210a ([pr/wipc-sep.100](https://github.com/ryanofsky/bitcoin/commits/pr/wipc-sep.100) -> [pr/wipc-sep.101](https://github.com/ryanofsky/bitcoin/commits/pr/wipc-sep.101), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/wipc-sep.100..pr/wipc-sep.101)) with suggestions from https://github.com/bitcoin/bitcoin/pull/10973#pullrequestreview-215654868.

> Incidentally, this github PR is pretty unwieldly

A downside of making a new PR is that comments that show up in the diffs will be gone. I think the diff comments are probably more useful than other historical comments which appear above, which I think no one will look at unless they are counting acks. (And if anyone is counting ACKs, there is only one so far from Marco in https://github.com/bitcoin/bitcoin/pull/10973#issuecomment-470238555 ([pr/wipc-sep.97](https://github.com/ryanofsky/bitcoin/commits/pr/wipc-sep.97)) before some recent cleanups.

jamesob reviewed (approved):
utACK https://github.com/bitcoin/bitcoin/commit/1a1036a5b3bc48715a5f3955e87203287c37210a

Changes look fine to me - feel free the ignore the nit. I'll do some manual testing of this tomorrow.

ryanofsky commented on src/wallet/wallet.h:
```diff
@@ -1220,6 +1224,8 @@ class CWallet final : public CCryptoKeyStore, public CValidationInterface
     /** Add a KeyOriginInfo to the wallet */
     bool AddKeyOrigin(const CPubKey& pubkey, const KeyOriginInfo& info);
+
+    friend struct WalletTestingSetup;
```
It's needed to avoid "error: cannot cast 'CWallet' to its private base class 'interfaces::Chain::Notifications'" in the WalletTestingSetup constructor:

https://github.com/bitcoin/bitcoin/blob/0ca5446db3cda7f911e911a094482311f6a71221/src/wallet/test/wallet_test_fixture.cpp#L16

because the relevant base class is private:

https://github.com/bitcoin/bitcoin/blob/0ca5446db3cda7f911e911a094482311f6a71221/src/wallet/wallet.h#L641

jnewbery reviewed (commented):
I've reviewed:

 - [x] 0ca5446db3cda7f911e911a094482311f6a71221 (_Remove use CValidationInterface in wallet code_) **utACK**
- [ ] f157e7a73ba746c38c5437540806151fbfa4b286 (_Remove use of CRPCTable::appendCommand in wallet code_)
- [x] 0bafbfcdd4b7d2393dc3e8a40b4ba7632117463a (_Remove use of CCoinsViewMemPool::GetCoin in wallet code_) and **utACK**
- [ ] 1a1036a5b3bc48715a5f3955e87203287c37210a (_Remove remaining wallet accesses to node globals_)

I found the _Remove use of CRPCTable::appendCommand in wallet code_ quite hard to review, probably due to my C++ ineptitude. I certainly wouldn't want to hold up this PR because I'm not able to fully work out what's going on in that commit. I think it could be made marginally simpler by removing support for multiple clients (https://github.com/bitcoin/bitcoin/pull/10973#discussion_r186538449), but Russ points out that most of the complexity is elsewhere. I also think some additional commenting in the `RpcForwarder` class could help.

Russ is going to look at moving more code into the RPC server to simplify that commit, but again, I don't want to hold this up because of my inability to review. If other reviewers are happy to ACK this, then it should stay as it is.

ryanofsky reviewed (commented):
Updated 1a1036a5b3bc48715a5f3955e87203287c37210a -> bffd676263953c662245be82c13a06c9bf5fbb20 ([pr/wipc-sep.101](https://github.com/ryanofsky/bitcoin/commits/pr/wipc-sep.101) -> [pr/wipc-sep.102](https://github.com/ryanofsky/bitcoin/commits/pr/wipc-sep.102), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/wipc-sep.101..pr/wipc-sep.102)) with John's suggestions.
Updated bffd676263953c662245be82c13a06c9bf5fbb20 -> d358466de15ef29c1d2bccb9aebab360d574d1d0 ([pr/wipc-sep.102](https://github.com/ryanofsky/bitcoin/commits/pr/wipc-sep.102) -> [pr/wipc-sep.103](https://github.com/ryanofsky/bitcoin/commits/pr/wipc-sep.103), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/wipc-sep.102..pr/wipc-sep.103)) renaming new src/node/coin.h file to be consistent with existing src/node/transaction.h file.


> I found the Remove use of CRPCTable::appendCommand in wallet code [f157e7a73ba746c38c5437540806151fbfa4b286] quite hard to review

New version of this commit (4e4d9e9f85eaf9c3bec48559bd4cad3e8a9333ca) should be a lot simpler. I was too reluctant before to change rpc server code. Should be much more straightforward now.

jnewbery commented:
utACK d358466de15ef29c1d2bccb9aebab360d574d1d0. This version is much clearer to me than the previous branch. Thank you!

I still think that handling multiple clients is an unnecessary complication for this PR. The `RPC_WALLET_NOT_FOUND` catch/throw in the `RpcHandlerImpl` actor and the `last_handler` bool passed to the actor seem particularly strange to me.

That said, I've reviewed the code and it all seems fine. I wouldn't want to hold this PR up because of my own aesthetic inclinations.

Very nice work Russ. Thanks for seeing this through!

MeshCollider commented:
I think this is ready 🎉

ryanofsky commented:
re: https://github.com/bitcoin/bitcoin/pull/10973#pullrequestreview-213000398

> Instead, I'd like to just add a new libbitcoin_node.a library, analogous to the existing libbitcoin_wallet.a library, and move symbol definitions like cs_main pcoinsTip chainActive there that wallet and gui code shouldn't access

To follow up, I did this in #15638 and #15639, only using the existing `libbitcoin_server.a` library for simplicity instead of adding a new `libbitcoin_node.a` library.
